### ISSUE TITLE
Electron 33 broken

### ISSUE THREAD

**Issue by danielweck** [nodejs/nan#978]
The latest NAN release works fine in Electron 32,  but now fails with:

```
In file included from ../node_modules/nan/nan.h:3045:
../node_modules/nan/nan_scriptorigin.h:40:7: error: no matching constructor for initialization of 'v8::ScriptOrigin'
   40 |       v8::ScriptOrigin(v8::Isolate::GetCurrent()
      |       ^                ~~~~~~~~~~~~~~~~~~~~~~~~~
   41 |                    , name
      |                    ~~~~~~
   42 |                    , To<int32_t>(line).FromMaybe(0)) {}
../node_modules/nan/nan_scriptorigin.h:47:7: error: no matching constructor for initialization of 'v8::ScriptOrigin'
   47 |       v8::ScriptOrigin(v8::Isolate::GetCurrent()
      |       ^                ~~~~~~~~~~~~~~~~~~~~~~~~~
   48 |                    , name
      |                    ~~~~~~
   49 |                    , To<int32_t>(line).FromMaybe(0)
      |                    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   50 |                    , To<int32_t>(column).FromMaybe(0)) {}

../node_modules/nan/nan_scriptorigin.h:36:7: error: no matching constructor for initialization of 'v8::ScriptOrigin'
   36 |       v8::ScriptOrigin(v8::Isolate::GetCurrent(), name) {}
      |       ^                ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/qwe/Library/Caches/node-gyp/33.0.0/include/node/v8-message.h:64:13: note: candidate constructor not viable: no known conversion from 'Isolate *' to 'Local<Value>' for 1st argument
   64 |   V8_INLINE ScriptOrigin(Local<Value> resource_name,
      |             ^            ~~~~~~~~~~~~~~~~~~~~~~~~~~
/Users/qwe/Library/Caches/node-gyp/33.0.0/include/node/v8-message.h:62:17: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 2 were provided
   62 | class V8_EXPORT ScriptOrigin {
      |                 ^~~~~~~~~~~~
/Users/qwe/Library/Caches/node-gyp/33.0.0/include/node/v8-message.h:62:17: note: candidate constructor (the implicit move constructor) not viable: requires 1 argument, but 2 were provided
   62 | class V8_EXPORT ScriptOrigin {
      |                 ^~~~~~~~~~~~
```

**Comment by danielweck** [comment]
This follows the fix for Electron v32:

* https://github.com/nodejs/nan/pull/976
* https://github.com/nodejs/nan/pull/977


See: https://github.com/nodejs/nan/issues/973#issuecomment-2404902320

**Comment by danielweck** [comment]
Electron 33 ships with Node 20.18.0:

https://github.com/electron/electron/releases/tag/v33.0.0

**Comment by agracio** [comment]
Electron 33 moved to v8 13 while Node.js main is currently on v8 12.9, perhaps someone can create a working branch that addresses the problem?
Unless by some miracle Node.js moves to v8 13 for its v23 release I do not think `nan` will address the issue themselves.

**Comment by danielweck** [comment]
Great point @agracio ... pinging @toshiyuki-rs @oc-soft @kkoopa @bugsounet

**Comment by agracio** [comment]
You should not ping `nan` maintainers as they have been quick to point out that they do not and have no intention to support Electron.

**Comment by danielweck** [comment]
Apologies, I thought *you* were the NAN maintainer :) 👍 (very quick response indeed, thank you)

Well, my workaround is to comment `#include nan_scriptorigin.h` in `nan.h`:

https://github.com/nodejs/nan/blob/f5921c7eb97dacb926485d7084ed475963773d07/nan.h#L3045C15-L3045C22

(I do this with `sed` from my `package.json` `postinstall`)

Closing the issue now.

**Comment by kkoopa** [comment]
It is probably this commit. https://github.com/nodejs/nan/pull/974/commits/0151569fd8ce287d675dafd3dac3eeecb59323b3 I will get around to it at some future time.

**Comment by agracio** [comment]
Thanks!

**Comment by agracio** [comment]
@danielweck could you share your package.json script with postinstall

**Comment by kkoopa** [comment]
And can someone with Electron 33 confirm that applying https://github.com/nodejs/nan/pull/974/commits/0151569fd8ce287d675dafd3dac3eeecb59323b3 resolves the issue?

**Comment by agracio** [comment]
Will do but give me about 30+ minutes.

**Comment by danielweck** [comment]
> @danielweck could you share your package.json script with postinstall

Not pretty, quick and dirty, but works in MacOS and Linux:

`"postinstall": "sed 's/^#include .nan_scriptorigin\\.h./\\/\\/ #include nan_scriptorigin.h/' ./node_modules/nan/nan.h > ./node_modules/nan/nan.h.new && mv ./node_modules/nan/nan.h.new ./node_modules/nan/nan.h"`

**Comment by agracio** [comment]
> And can someone with Electron 33 confirm that applying [0151569](https://github.com/nodejs/nan/commit/0151569fd8ce287d675dafd3dac3eeecb59323b3) resolves the issue?

It absolutely does, tested Electron 29, 30, 31, 32 and 33 that are supported by nan 2.22.

**Comment by agracio** [comment]
@kkoopa So is it just a issue of merging that commit and everything will be working?

**Comment by ayZagen** [comment]
@danielweck this issue is still valid and should be open until a fix is landed

**Comment by agracio** [comment]
The issues is not present in nan v2.22.2 tested with electron 29-34 as well as 35 beta and latest 36 nightly.

**Comment by agracio** [comment]
While the issue is solved in `nan v2.22.2` postinstall overrides can easily be applied for sub dependencies. 
It is no longer needed but I can provide a sample .js script from my repo if anyone is interested.

**Comment by oalfroukh** [comment]
> While the issue is solved in `nan v2.22.2` postinstall overrides can be easily applied for sub dependencies. Here is a sample script from my repo:
> 
> `updatenan.js`
> 
> const spawn = require('child_process').spawn
> const existsSync = require('fs').existsSync
> const nanPath = existsSync('./node_modules/nan/nan.h') ? './node_modules/nan/nan.h' : '../nan/nan.h'
> 
> if(existsSync(nanPath)){
>     if (process.platform === 'win32') {
>         spawn('powershell', ['-Command', `(Get-Content -Raw ${nanPath}) -replace '#include \"nan_scriptorigin.h\"', '// #include \"nan_scriptorigin.h\"' | Out-File -Encoding Utf8 ${nanPath}`], { stdio: 'inherit' })
>     } else {
>         spawn('sed', ['-i', '-e', 's/^#include .nan_scriptorigin\\.h./\\/\\/ #include nan_scriptorigin.h/', nanPath], { stdio: 'inherit' })
>     }
> 
> }

Also just adding `overrides` to the `package.json` should fix the issue for sub dependencies instead of using scripts 

```
,
  "overrides": {
    "nan": "2.22.2"
  }
```

**Comment by agracio** [comment]
I removed that script from my repo since the release of `nan v2.22.2` just wanted to show a PoC of install/postinstall script that works for sub dependencies. Cannot even take a credit for it as it was a PR from one of the repo contributors.
